include/gazebo_vision_plugin.h

```diff
@@ -91,13 +91,16 @@ class GAZEBO_VISIBLE VisionPlugin : public ModelPlugin
 
   transport::NodePtr _nh;
   transport::PublisherPtr _pub_odom;
-#if BUILD_ROS_INTERFACE == 1
+#if BUILD_ROS_INTERFACE == ON
   void queueThread();
   std::string _ros_sub_topic;
   ros::CallbackQueue _ros_queue;
   std::unique_ptr<ros::NodeHandle> _ros_node;
   ros::Subscriber _ros_sub;
   std::thread _ros_queue_thread;
+  bool _enable_ros_odom;
+  static constexpr auto kDefaultRosOdomTopic = "/svo/pose";
```

Comment:
> This should be a common topic name. Example: "/visual/odometry". Then one should do a topic remap if needed.

Agree.

> Also, options should be added for the topic msg type. Currently you just give the option for geometry_msgs::PoseWithCovarianceStamped but it should be open for geometry_msgs::PoseStamped and nav_msgs::Odometry

Do you mean overloaded callback for such topics message types?
The problem here is how to test these cases if only one type of message topic is posted by external VO.
For the geometry_msgs::PoseStamped it is possible to write the node that converts to it from geometry_msgs::PoseWithCovarianceStamped.
But i don't have a sensor that could completely fill nav_msgs::Odometry.

Reply:
What I mean is that you should have a callback for those msgs, even if you don't have a framework that does send those msgs. There are a lot out there that are able to do it, so this becomes more flexible. Check for example what's done in MAVROS. Basically you add a callback for each topic, and then the data is handled on the same method, basically filling the fields of the `odometry` Gazebo msg that it is able to so. If the callback is for `navigation_msgs/Odometry.msg` type, then you will be able to fill more that. If it is just `Pose` data, then you will fill less data. Will be better on that case to fill with NaNs, which in the case of the ODOMETRY Mavlink msg, for now, would mean 0's (Have to update the msg spec so we are able to state which data is valid or not).

Reply:
I've added handlers, but not subscribers. The reason is that if multiple handlers with different arguments are subscribed to the one topic then we will get ROS error. I leave this problem for now.

Reply:
So why not different subscribers? You do need different handler per subscriber.

Reply:
Im not quite sure. Suppose we have Odometry handler and PoseStamped handler and they both are subscribed to "/visual/odometry" topic. If VO framework advertises topic with Odometry type, then PoseStamped handler crashes because it has different type. But i may be wrong i will check it.

Reply:
```c++
if(sdf->HasElement("rosPoseSubTopic")) {
    _ros_pose_sub_topic = sdf->GetElement("rosPoseSubTopic")->Get<std::string>();
  } else {
    _ros_pose_sub_topic = kDefaultRosPoseTopic;
    gzwarn << "[gazebo_vision_plugin] Using default ROS PoseStamped subscription to topic " << _ros_pose_sub_topic << "\n";
  }

if(sdf->HasElement("rosPoseCovSubTopic")) {
    _ros_posecov_sub_topic = sdf->GetElement("rosPoseCovSubTopic")->Get<std::string>();
  } else {
    _ros_posecov_sub_topic = kDefaultRosPoseCovTopic;
    gzwarn << "[gazebo_vision_plugin] Using default ROS PoseWithCovarianceStamped subscription to topic " << _ros_posecov_sub_topic << "\n";
  }

if(sdf->HasElement("rosOdomSubTopic")) {
    _ros_odom_sub_topic = sdf->GetElement("rosOdomSubTopic")->Get<std::string>();
  } else {
    _ros_posecov_sub_topic = kDefaultRosOdomTopic;
    gzwarn << "[gazebo_vision_plugin] Using default ROS Odometry subscription to topic " << _ros_odom_sub_topic << "\n";
  }

(...)

ros::SubscribeOptions so_pose = ros::SubscribeOptions::create<geometry_msgs::PoseStamped>(_ros_pose_sub_topic, 1, boost::bind(&VisionPlugin::rosPoseCallBack, this, _1), ros::VoidPtr(), &this->_ros_queue);
ros::SubscribeOptions so_posecov = ros::SubscribeOptions::create<geometry_msgs::PoseWithCovarianceStamped>(_ros_posecov_sub_topic, 1, boost::bind(&VisionPlugin::rosPoseCovCallBack, this, _1), ros::VoidPtr(), &this->_ros_queue);
ros::SubscribeOptions so_odom = ros::SubscribeOptions::create<navigation_msgs::Odometry>(_ros_odom_sub_topic, 1, boost::bind(&VisionPlugin::rosOdomCallBack, this, _1), ros::VoidPtr(), &this->_ros_queue);

this->_ros_pose_sub = this->_ros_node->subscribe(so_pose);
this->_ros_posecov_sub = this->_ros_node->subscribe(so_posecov);
this->_ros_odom_sub = this->_ros_node->subscribe(so_odom);

(...)
```

Reply:
I see your point. In this case problem disappears of course.